core/script_language.cpp

```diff
@@ -360,6 +396,190 @@ void ScriptDebugger::clear_breakpoints() {
+	watch.locals = Vector<String>();
+	watch.script_instance = NULL;
+
+	set_watch_expression(watches.size() - 1, p_expression);
+}
+
+void ScriptDebugger::set_watch_expression(int p_index, const String &p_expression) {
+	ERR_FAIL_INDEX(p_index, watches.size());
+
+	WatchData &watch = watches.write[p_index];
+	if (watch.expression->parse(p_expression, Vector<String>()) != OK) {
+		watch.error = true;
+	}
+}
+
+void ScriptDebugger::set_watch_tracking(int p_index, bool p_is_tracking) {
+	ERR_FAIL_INDEX(p_index, watches.size());
+	WatchData &watch = watches.write[p_index];
+
+	watch.tracking = p_is_tracking;
+	if (p_is_tracking) {
+
+		ScriptLanguage *lang = get_break_language();
+		const String &function_name = lang->debug_get_stack_level_function(0);
+		Expression::Context context = lang->debug_get_stack_level_expression_context(0, watch.expression->get_source_string());
+
+		watch.function_name = function_name;
+		watch.locals.append_array(context.input_names);
+		watch.script_instance = lang->debug_get_stack_level_instance(0);
+	}
+}
+
+void ScriptDebugger::set_watch_lock(int p_stack_level, int p_index, bool p_is_locked) {
+	ERR_FAIL_INDEX(p_index, watches.size());
+	WatchData &watch = watches.write[p_index];
+	ScriptLanguage *lang = get_break_language();
+
+	if (lang->debug_get_stack_level_count() != 0) {
+		const String &expression_string = watch.expression->get_source_string();
+		Expression::Context context = lang->debug_get_stack_level_expression_context(p_stack_level, expression_string);
+
+		watch.function_name = lang->debug_get_stack_level_function(p_stack_level);
+		watch.script_instance = lang->debug_get_stack_level_instance(p_stack_level);
+		watch.locals = context.input_names;
+	}
+	watch.locked = p_is_locked;
+}
+
+void ScriptDebugger::set_watch_dirty(int p_index, bool p_dirty) {
+	ERR_FAIL_INDEX(p_index, watches.size());
+	watches.write[p_index].dirty = p_dirty;
+}
+
+void ScriptDebugger::remove_watch(int p_index) {
+	ERR_FAIL_INDEX(p_index, watches.size());
+
+	memdelete(watches[p_index].expression);
+	watches.remove(p_index);
+}
+
+void ScriptDebugger::clear_watches() {
+	const int size = watches.size();
+
+	for (int i = 0; i < size; i++) {
+		memdelete(watches[i].expression);
+	}
+	watches.clear();
+}
+
+Variant ScriptDebugger::get_watch_result(int p_index) const {
+	ERR_FAIL_INDEX_V(p_index, watches.size(), Variant());
+
+	return watches[p_index].result_cache;
+}
+
+bool ScriptDebugger::watch_has_error(int p_index) const {
+	ERR_FAIL_INDEX_V(p_index, watches.size(), false);
+
+	return watches[p_index].error;
+}
+
+bool ScriptDebugger::watch_is_dirty(int p_index) const {
+	ERR_FAIL_INDEX_V(p_index, watches.size(), false);
+
+	return watches[p_index].dirty;
+}
+
+String ScriptDebugger::get_watch_error_text(int p_index) const {
+	ERR_FAIL_INDEX_V(p_index, watches.size(), String());
+
+	return watches[p_index].expression->get_error_text();
+}
+
+bool ScriptDebugger::evaluate_watches(int p_stack_level, int p_watch) {
```

Comment:
Some feedback:
1. This function will be called for every line of script code being run, and it will make execution considerably slower so make sure nothing is run if no watches exist by using some sort of inline static bool to check watches.
2. This function can be called from multiple threads, make sure to add a mutex to it if it actually runs
3. You will need to check if waches use stack variables from the current function, if so they will need to be erased once function exits.